### ISSUE TITLE
systemctl restart systemd-nspawn@<name> consistently fails

### ISSUE THREAD

**Issue by arianvp** [systemd/systemd#14961]
**systemd version the issue has been seen with**
> v243

**Used distribution**
 > NixOS

**Expected behaviour you didn't see**
 `systemctl restart systemd-nspawn@test` cleanly restarts the container

**Unexpected behaviour you saw**
It fails.

It then stops the container anyway, so the next `systemctl restart` works again as it merely starts
```
$ sudo systemctl restart systemd-nspawn@test1

$ sudo systemctl restart systemd-nspawn@test1
Job for systemd-nspawn@test1.service failed because the control process exited with error code.
See "systemctl status systemd-nspawn@test1.service" and "journalctl -xe" for details.

[arian@t490s:~/Projects/nixos-stuff]$ sudo systemctl restart systemd-nspawn@test1

[arian@t490s:~/Projects/nixos-stuff]$ sudo systemctl restart systemd-nspawn@test1
Job for systemd-nspawn@test1.service failed because the control process exited with error code.
See "systemctl status systemd-nspawn@test1.service" and "journalctl -xe" for details.


```

Logs from `systemd-nspawn@test.service`:
```
Feb 27 16:04:33 t490s systemd-nspawn[27421]: /etc/localtime does not point into /usr/share/zoneinfo/, not updating container timezone.
Feb 27 16:04:33 t490s systemd-nspawn[27421]: Failed to register machine: Machine 'test1' already exists
Feb 27 16:04:33 t490s systemd-nspawn[27421]: Parent died too early
Feb 27 16:04:34 t490s systemd[1]: systemd-nspawn@test1.service: Main process exited, code=exited, status=1/FAILURE
Feb 27 16:04:34 t490s systemd[1]: systemd-nspawn@test1.service: Failed with result 'exit-code'.
Feb 27 16:04:34 t490s systemd[1]: Failed to start Container test1.
-- Subject: Unit systemd-nspawn@test1.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@test1.service has failed.
-- 
-- The result is RESULT.
```

Logs from `systemd-machined`:
```
-- The virtual machine test1 with its leader PID 23266 has been
-- started is now ready to use.
Feb 27 15:55:32 t490s systemd-machined[1195]: Failed to stop machine scope: Transaction for systemd-nspawn@test1.service/stop is destructive (systemd-nspawn@test1.service has 'restart' job queued, but 'stop' is included in transaction).
Feb 27 15:55:32 t490s systemd-machined[1195]: Failed to drop reference to machine scope, ignoring: Unit has not been referenced yet.
Feb 27 15:55:32 t490s systemd-machined[1195]: Machine test1 terminated.
-- Subject: A virtual machine or container has been terminated
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- The virtual machine test1 with its leader PID 23266 has been
-- shut down.
Feb 27 15:55:35 t490s systemd-machined[1195]: New machine test1.
-- Subject: A virtual machine or container has been started
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- The virtual machine test1 with its leader PID 23577 has been
-- started is now ready to use.
Feb 27 15:55:37 t490s systemd-machined[1195]: Failed to stop machine scope: Transaction for systemd-nspawn@test1.service/stop is destructive (systemd-nspawn@test1.service has 'restart' job queued, but 'stop' is included in transaction).
Feb 27 15:55:37 t490s systemd-machined[1195]: Failed to drop reference to machine scope, ignoring: Unit has not been referenced yet.
Feb 27 15:55:38 t490s systemd-machined[1195]: Machine test1 terminated.
-- Subject: A virtual machine or container has been terminated
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- The virtual machine test1 with its leader PID 23577 has been
-- shut down.
```

**Steps to reproduce the problem**

`test1.nspawn`:

```
[Exec]
Boot=false
# Because nixos init system is not in /sbin/init ,we have to set KillSignal manually and Boot=false
KillSignal=SIGRTMIN+3
Parameters=/nix/store/9n1bbglvg22v4c9ah9gdgcfn8nxnkzvl-nixos-system-test1-19.09pre-git/init
PrivateUsers=no


[Files]
BindReadOnly=/nix/store
BindReadOnly=/nix/var/nix/db
BindReadOnly=/nix/var/nix/daemon-socket


[Network]
Zone=nixos
```

**Comment by arianvp** [comment]
`systemctl stop` followed by `systemctl start` works but I get the following warning from `systemd-machined`:

```
Feb 27 16:00:11 t490s systemd-machined[1195]: Failed to drop reference to machine scope, ignoring: Unit has not been referenced yet.
```

**Comment by arianvp** [comment]
Ok it seems `restart` consistently fails with me.  but it's just the second time it succeeds as the container isn't running then anymore and it just a `start`

Logs from `systemd-nspawn@test1` when issuing the restart that fails:


```
eb 27 16:08:27 t490s systemd-nspawn[28422]: Trying to halt container. Send SIGTERM again to trigger immediate termination.
Feb 27 16:08:27 t490s systemd[1]: Stopping Container test1...
-- Subject: Unit systemd-nspawn@test1.service has begun shutting down
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@test1.service has begun shutting down.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Removed slice system-getty.slice.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Login Prompts.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Host and Network Name Lookups.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Timers.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Daily Cleanup of Temporary Directories.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Console Getty...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping D-Bus System Message Bus...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Nginx Web Server...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Login Service...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Console Getty.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Permit User Sessions...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped D-Bus System Message Bus.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Wait for Network to be Configured.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Permit User Sessions.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Remote File Systems.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Login Service.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target User and Group Name Lookups.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Name Service Cache Daemon...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Nginx Web Server.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Network.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target All Network Interfaces (deprecated).
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Extra networking commands..
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Networking Setup.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Network Name Resolution...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Name Service Cache Daemon.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Basic System.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Paths.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Dispatch Password Requests to Console Directory Watch.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Forward Password Requests to Wall Directory Watch.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Slices.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Removed slice User and Session Slice.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Sockets.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Closed D-Bus System Message Bus Socket.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target System Initialization.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Update UTMP about System Boot/Shutdown...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Network Name Resolution.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Stopping Network Service...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Update UTMP about System Boot/Shutdown.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Create Volatile Files and Directories.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Local File Systems.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Unmounting /nix/var/nix/daemon-socket...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Unmounting /nix/var/nix/db...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Unmounting /run/keys...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Unmounting /run/systemd/nspawn/incoming...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Unmounting /run/wrappers...
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Unmounting /tmp...
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Network Service.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Network (Pre).
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Apply Kernel Variables.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Unmounted /nix/var/nix/daemon-socket.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Unmounted /nix/var/nix/db.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Unmounted /run/systemd/nspawn/incoming.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Unmounted /run/keys.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Unmounted /tmp.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Unmounted /run/wrappers.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Local File Systems (Pre).
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped target Swap.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Reached target Unmount All Filesystems.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Stopped Create Static Device Nodes in /dev.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Reached target Shutdown.
Feb 27 16:08:27 t490s systemd-nspawn[28422]: [  OK  ] Reached target Final Step.
Feb 27 16:08:27 t490s systemd-nspawn[28422]:          Starting Halt...
Feb 27 16:08:27 t490s systemd[1]: systemd-nspawn@test1.service: Succeeded.
Feb 27 16:08:27 t490s systemd[1]: Stopped Container test1.
-- Subject: Unit systemd-nspawn@test1.service has finished shutting down
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@test1.service has finished shutting down.
Feb 27 16:08:27 t490s systemd[1]: systemd-nspawn@test1.service: Consumed 1.511s CPU time, received 0B IP traffic, sent 1.0K IP traffic.
-- Subject: Resources consumed by unit runtime
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- The unit systemd-nspawn@test1.service completed and consumed the indicated resources.
Feb 27 16:08:27 t490s systemd[1]: Starting Container test1...
-- Subject: Unit systemd-nspawn@test1.service has begun start-up
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@test1.service has begun starting up.
Feb 27 16:08:27 t490s systemd[28714]: systemd-nspawn@test1.service: Executable /sbin/modprobe missing, skipping: No such file or directory
-- Subject: Process /sbin/modprobe could not be executed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- The process /sbin/modprobe could not be executed and failed.
-- 
-- The error number returned by this process is ERRNO.
Feb 27 16:08:27 t490s systemd-nspawn[28718]: /etc/localtime does not point into /usr/share/zoneinfo/, not updating container timezone.
Feb 27 16:08:27 t490s systemd-nspawn[28718]: Failed to register machine: Machine 'test1' already exists
Feb 27 16:08:27 t490s systemd[1]: systemd-nspawn@test1.service: Main process exited, code=exited, status=1/FAILURE
Feb 27 16:08:27 t490s systemd[1]: systemd-nspawn@test1.service: Failed with result 'exit-code'.
Feb 27 16:08:27 t490s systemd[1]: Failed to start Container test1.
-- Subject: Unit systemd-nspawn@test1.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@test1.service has failed.
-- 
-- The result is RESULT.
```

**Comment by flokli** [comment]
Can you please check if this is a NixOS-specific issue, or if you run into the same problem with other distros, like Fedora Rawhide?

**Comment by DaanDeMeyer** [comment]
Can you try v244? I remember one of the last remaining causes of this bug was fixed in v244.

**Comment by poettering** [comment]
Yeah, this looks like what was already fixed with eec12b7756e2cada16e3b8a9c02f9eaf62df5409 i.e 244.

Let's close this here. If you manage to reproduce this in current systemd, please reopen
